### PR TITLE
Secondary index: Respect limit and page size, do not skip clustering ranges after short reads

### LINKED ISSUE #7432: Some queries with secondary indexes return too large pages
```
cqlsh> CREATE KEYSPACE ks WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1};
cqlsh> CREATE TABLE ks.t2(pk int, ck1 int, ck2 int, PRIMARY KEY(pk, ck1, ck2));
cqlsh> CREATE INDEX ON ks.t2(ck1);
cqlsh> INSERT INTO ks.t2(pk, ck1, ck2) VALUES (1, 1, 1);
cqlsh> INSERT INTO ks.t2(pk, ck1, ck2) VALUES (1, 1, 2);
cqlsh> INSERT INTO ks.t2(pk, ck1, ck2) VALUES (1, 1, 3);
cqlsh> INSERT INTO ks.t2(pk, ck1, ck2) VALUES (1, 1, 4);
cqlsh> INSERT INTO ks.t2(pk, ck1, ck2) VALUES (2, 1, 1);
cqlsh> INSERT INTO ks.t2(pk, ck1, ck2) VALUES (2, 1, 2);
cqlsh> INSERT INTO ks.t2(pk, ck1, ck2) VALUES (3, 1, 1);
cqlsh> INSERT INTO ks.t2(pk, ck1, ck2) VALUES (3, 1, 2);
cqlsh> PAGING 1
Page size: 1
cqlsh> SELECT pk FROM ks.t2 WHERE ck1 = 1 GROUP BY pk;

 pk
----
  3
  1
  2

(3 rows)
```

```
cqlsh> CREATE KEYSPACE ks WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1};
cqlsh> CREATE TABLE ks.t3(pk1 int, pk2 int, ck int, PRIMARY KEY((pk1, pk2), ck));
cqlsh> CREATE INDEX ON ks.t3(pk2);
cqlsh> INSERT INTO ks.t3(pk1, pk2, ck) VALUES (1, 1, 1);
cqlsh> INSERT INTO ks.t3(pk1, pk2, ck) VALUES (1, 1, 2);
cqlsh> INSERT INTO ks.t3(pk1, pk2, ck) VALUES (1, 1, 3);
cqlsh> INSERT INTO ks.t3(pk1, pk2, ck) VALUES (1, 1, 4);
cqlsh> INSERT INTO ks.t3(pk1, pk2, ck) VALUES (1, 1, 5);
cqlsh> PAGING 1
Page size: 1
cqlsh> SELECT * FROM ks.t3 WHERE pk2 = 1;

 pk1 | pk2 | ck
-----+-----+----
   1 |   1 |  1
   1 |   1 |  2
   1 |   1 |  3
   1 |   1 |  4
   1 |   1 |  5

---MORE---
(5 rows)
```

In both of this queries, a too large page is returned. In the first example, a single page of size 3 is returned. In the second example a page of size 5 is returned and an empty page afterwards. Cassandra 3.11.8 correctly pages both of these queries. Please note that the first example depends on my fixes for issue #7355 (however, before those fixes a single too large page was also returned).

### LINKED ISSUE #22158: Secondary index query may exceed the LIMIT
Assume we create a secondary index on the first column of a clustering key. When running a point query on the indexed column and we use a limit, the limit seems to be exceeded when the result spans across more than one partitions. Here's an example:
```
cqlsh> CREATE KEYSPACE ks WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1};
cqlsh> CREATE TABLE ks.t1 (col1 int, col2 int, col3 int, PRIMARY KEY (col1, col2, col3));
cqlsh> CREATE INDEX on ks.t1 (col2);
cqlsh> INSERT INTO ks.t1 (col1, col2, col3) VALUES (1, 1, 1);
cqlsh> INSERT INTO ks.t1 (col1, col2, col3) VALUES (1, 1, 2);
cqlsh> INSERT INTO ks.t1 (col1, col2, col3) VALUES (2, 1, 1);
cqlsh> INSERT INTO ks.t1 (col1, col2, col3) VALUES (2, 1, 2);
cqlsh> SELECT * FROM ks.t1 WHERE col2=1 LIMIT 1;

 col1 | col2 | col3
------+------+------
    1 |    1 |    1

(1 rows)
cqlsh> SELECT * FROM ks.t1 WHERE col2=1 LIMIT 2;

 col1 | col2 | col3
------+------+------
    1 |    1 |    1
    1 |    1 |    2

(2 rows)
cqlsh> SELECT * FROM ks.t1 WHERE col2=1 LIMIT 3;

 col1 | col2 | col3
------+------+------
    1 |    1 |    1
    1 |    1 |    2
    2 |    1 |    1
    2 |    1 |    2

---MORE---
(4 rows)
```
The last query returned 4 rows instead of 3. Also, it was paged, although the paging size is the default (100).

Related issues:
* https://github.com/scylladb/scylladb/issues/20660

### LINKED ISSUE #25839: Secondary index: large rows + paging cause incomplete results
Let's assume we have a table with:
* multi-row partitions
* a regular column storing large text data
* a secondary index on a partition key column or the first clustering key column (i.e., a column that causes reading whole partitions or partition slices from the base table)

If we run a paged query on the indexed column and the returned page is short, then the next page will skip all the remaining rows from the first page's last partition (as indicated by the paging state). 

A cqlpy reproducer:
```python
def test_short_read(cql, test_keyspace):
    page_memory_limit = 1024 * 1024  # 1MiB (taken from `result_memory_accounter::maximum_result_size`)
    row_size = page_memory_limit // 2  # will cause short read after 2 rows
    with new_test_table(cql, test_keyspace, 'pk int, ck1 int, ck2 int, data text, primary key (pk, ck1, ck2)') as table:
        cql.execute(f'CREATE INDEX ON {table}(ck1)')
        stmt = cql.prepare(f'INSERT INTO {table} (pk, ck1, ck2, data) VALUES (?, ?, ?, ?)')
        cql.execute(stmt, [1, 1, 1, 'A' * row_size])
        cql.execute(stmt, [1, 1, 2, 'B' * row_size])
        cql.execute(stmt, [1, 1, 3, 'C' * row_size])
        cql.execute(stmt, [2, 1, 1, 'E' * row_size])
        stmt = SimpleStatement(f'SELECT pk, ck2, data FROM {table} WHERE ck1 = 1')

        # First page
        rs = cql.execute(stmt)
        page1_rows = [(col[0], col[1]) for col in rs.current_rows]
        assert page1_rows == [(1,1), (1,2)]
        assert rs.has_more_pages == True

        # Second page
        rs2 = cql.execute(stmt, paging_state=rs.paging_state)
        page2_rows = [(col[0], col[1]) for col in rs2.current_rows]
        assert rs2.has_more_pages == False
        assert page2_rows == [(1,3)]
```
Test result:
```
            assert rs2.has_more_pages == False
>           assert page2_rows == [(1,3)]
E           assert [(2, 1)] == [(1, 3)]
E             
E             At index 0 diff: (2, 1) != (1, 3)
E             Use -v to get more diff
```
In Cassandra 5, no rows are skipped and the whole result set is returned in a single page.

cc @nyh @ptrsmrn

### PR DESCRIPTION
**Pull request by ndragazis**
This PR attempts to strike three interrelated bugs at once:

* https://github.com/scylladb/scylladb/issues/25839
* https://github.com/scylladb/scylladb/issues/7432
* https://github.com/scylladb/scylladb/issues/22158

The PR starts by adding some additional test cases for #22158, and then fixes those issues incrementally as follows:

1. [Index view read] Restrict partition key deduplication across pages
    * Description: When a page stops in the midst of a partition (e.g., short read), do not skip this last partition from the next page, except for the very special case where the index is on a map column's values. Essentially, narrow down the skipping logic so that it applies only to the very special case that it was supposed to solve in the first place.
    * Effect: Fixes 25839 (except for a corner case with an index on map values of a static column; explained later in this cover letter). Required for 7432 and 22158.
2. [Index view read] Fix remaining rows when adjusting paging state
    * Description: When adjusting the index view's paging state after a base table read, do not bust the "remaining" field.
    * Effect: Required for 7432 and 22158.
3. [Base table read] Limit base table reads to match index view rows
    * Description: When reading partitions from the base table, apply an upper bound to the clustering ranges being read, so that the read stops at the same clustering key as the index view.
    * Effect: Fixes 7432 for the non-group-by case. Fixes 22158 for paged queries.
4. [Base table read] Fix limit for result merger in fast path
    * Description: Apply a stricter limit to the result merger that merges the results from the base table.
    * Effect: Fixes 22158 for the remaining cases (unpaged queries, queries on static columns).

State of issues after this PR (with references to cqlpy tests):
* #25839
   * Resolved when indexed column is not a map (`test_secondary_index::test_short_read`).
   * NOT resolved for index on map values of static column (`test_secondary_index::test_short_read_static_column_map_values`).
* #7432
   * Resolved for non-static non-group-by case (`test_secondary_index::test_index_paging_match_partition`).
   * NOT resolved for static case (`test_secondary_index::test_index_paging_static_column`).
   * NOT resolved for group-by case (`test_secondary_index::test_index_paging_group_by`).
* #22158
   * Fully resolved (`test_limit_partition`, `test_limit_partition_slice`, `test_limit_partition_slice_across_pages`, `test_limit_partition_slice_short_read`, `test_static_column_index_with_limit`).

Fixes #22158.
Refs #25839.
Refs #7432.

Intrusive changes, do not backport.

### CODE REVIEW

**Comment by nyh** [comment]
> Fixes #22158. Partially fixes #25839. Partially fixes #7432.

I believe that github's automation isn't as smart as you think, and when it sees "partially fixes #25839" it will just see the "fixes #25839" substring, and will close the issue. So please chose a different wording that doesn't include the magic substring.

Also, "partially fixes" is mysterious - what does it mean? How can you "partially" fix a bug? Do you mean the behavior is now slightly better but still wrong? Maybe you can explain exactly what the remaining situation of each issue is. After this pull request will be merged, you might also want to go into those two issues and leave a comment how this PR changed it and what problem still remains. Thanks.

**Comment by ndragazis** [comment]
> I believe that github's automation isn't as smart as you think, and when it sees "partially fixes #25839" it will just see the "fixes #25839" substring, and will close the issue. So please chose a different wording that doesn't include the magic substring.

Sigh, you're right.

> 
> Also, "partially fixes" is mysterious - what does it mean? How can you "partially" fix a bug? Do you mean the behavior is now slightly better but still wrong? Maybe you can explain exactly what the remaining situation of each issue is. After this pull request will be merged, you might also want to go into those two issues and leave a comment how this PR changed it and what problem still remains. Thanks.

I was hoping that "partially fixes" would be more intuitive than "refs".

#25839 is completely fixed except for a corner case where the indexed column is a static map. This is currently blocked on another bug (https://github.com/scylladb/scylladb/issues/16295), so I cannot even test this. I have added an xfail-ing test case.

#7432 has two different cases: with and without group by. This PR only fixes the former (and removes the xfail marker from the corresponding test). I haven't looked into the latter, but we have separate xfail-ing tests for that.

So, for both issues, we have xfail-ing tests for the missing pieces.

**Comment by mykaul** [comment]
> > Fixes #22158. Partially fixes #25839. Partially fixes #7432.
> 
> I believe that github's automation isn't as smart as you think, and when it sees "partially fixes #25839" it will just see the "fixes #25839" substring, and will close the issue. So please chose a different wording that doesn't include the magic substring.
> 
> Also, "partially fixes" is mysterious - what does it mean? How can you "partially" fix a bug? Do you mean the behavior is now slightly better but still wrong? Maybe you can explain exactly what the remaining situation of each issue is. After this pull request will be merged, you might also want to go into those two issues and leave a comment how this PR changed it and what problem still remains. Thanks.

We should use 'refs' for this.

**Comment by ndragazis** [comment]
> https://github.com/scylladb/scylladb/issues/25839 is completely fixed except for a corner case where the indexed column is a static map. This is currently blocked on another bug (https://github.com/scylladb/scylladb/issues/16295), so I cannot even test this. I have added an xfail-ing test case.
>
> https://github.com/scylladb/scylladb/issues/7432 has two different cases: with and without group by. This PR only fixes the former (and removes the xfail marker from the corresponding test). I haven't looked into the latter, but we have separate xfail-ing tests for that.
>
> So, for both issues, we have xfail-ing tests for the missing pieces.

The xfail-ing tests:
* `test_index_paging_static_column`
* `test_index_paging_group_by`
* `test_short_read_static_column_map_values` (added in this PR)

The first and third tests have the same root cause: if the indexed column is static, the index stores base table partition keys, not primary keys. This means that we cannot page through base table partitions because the paging state tracks the index view (i.e., base table partition keys), so the base clustering key is not recorded in the paging state.

The second test fails because the query scans all partitions internally and returns the whole result set as one page ([the code](https://github.com/scylladb/scylladb/blob/520cc0eeaa5002bbd78010818f5a13030757a3a4/cql3/statements/select_statement.cc#L1269-L1278)). Not sure why we do that. If I'm not mistaken, GROUP BY computes one row per partition, so paging could be implemented by tracking the number of partitions. Note that this also affects GROUP BY without a secondary index.

**Comment by nyh** [comment]
So @ndragazis please try to summarize shortly, in the commit message, what remains of the two issues you marked "refs", and please also go to these issues to explain in a comment there what remains of these issues. E.g., you wrote above that "https://github.com/scylladb/scylladb/issues/7432 has two different cases: with and without group by. This PR only fixes the former (and removes the xfail marker from the corresponding test).". The commit message can say something like this, and a comment in #7432 should perhaps mention that it only happens now without group by.
 
#7432 could be closed and replaced by a new issue outlining the remaining problem - but I don't think you need to waste your time doing this. As you said, we already have cqlpy tests for the actual remaining #7432 problems.

**Comment by ndragazis** [comment]
v2 Changelog:
* Explained in the cover letter how the state of the issues changes after this PR.
* Rebased to resolve conflicts (caused by the now-merged dependency: https://github.com/scylladb/scylladb/pull/25676).
* Renamed test `test_short_read_map_values` to `test_short_read_static_column_map_values`.
* Added some Refs in some commits so that it is more clear how they relate to each issue.

**Comment by ndragazis** [comment]
Conflicts resolved. Also https://github.com/scylladb/scylladb/pull/25843 has been merged, so I can now undraft this PR.

**Comment by ndragazis** [comment]
> #### Failed Tests (24/54270):
>
>  * [test_index_metrics.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21740/testReport/junit/cqlpy/test_secondary_index/release___test_py___release___test_index_metrics_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_index_metrics.release.1)
>  * [test_indexing_paging_and_aggregation.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21740/testReport/junit/cqlpy/test_secondary_index/release___test_py___release___test_indexing_paging_and_aggregation_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_indexing_paging_and_aggregation.release.1)
>  * [test_tombstone_gc_property.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21740/testReport/junit/cqlpy/test_secondary_index/release___test_py___release___test_tombstone_gc_property_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_property.release.1)
>  * [test_tombstone_gc_property_unnamed_index.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21740/testReport/junit/cqlpy/test_secondary_index/release___test_py___release___test_tombstone_gc_property_unnamed_index_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_property_unnamed_index.release.1)
>  * [test_index_metrics.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21740/testReport/junit/cqlpy/test_secondary_index/release___test_py___release___test_index_metrics_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_index_metrics.release.2)
>  * [test_indexing_paging_and_aggregation.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21740/testReport/junit/cqlpy/test_secondary_index/release___test_py___release___test_indexing_paging_and_aggregation_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_indexing_paging_and_aggregation.release.2)
>  * [test_tombstone_gc_property.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21740/testReport/junit/cqlpy/test_secondary_index/release___test_py___release___test_tombstone_gc_property_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_property.release.2)
>  * [test_tombstone_gc_property_unnamed_index.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21740/testReport/junit/cqlpy/test_secondary_index/release___test_py___release___test_tombstone_gc_property_unnamed_index_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_property_unnamed_index.release.2)
>  * [test_index_metrics.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21740/testReport/junit/cqlpy/test_secondary_index/release___test_py___release___test_index_metrics_release_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_index_metrics.release.3)
>  * [test_indexing_paging_and_aggregation.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21740/testReport/junit/cqlpy/test_secondary_index/release___test_py___release___test_indexing_paging_and_aggregation_release_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_indexing_paging_and_aggregation.release.3)

All these tests fail because they cannot connect to the Scylla node:
> ```failed on setup with "requests.exceptions.ConnectionError: HTTPConnectionPool(host='127.179.47.56', port=10000): Max retries exceeded with url: /system/log?message=test/cqlpy%3A%20Starting%20test_secondary_index.py%3A%3Atest_index_metrics.release.1&level=info (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f0788296be0>: Failed to establish a new connection: [Errno 111] Connection refused'))"```

The node crashed while running the xfailing test `test_short_read_static_column_map_values` that is introduced in this PR. Specifically, an INSERT statement triggered an `std::bad_alloc`, but because tests run Scylla with `--abort-on-seastar-bad-alloc`, Scylla aborted.
```
WARN  2025-11-13 11:56:36,504 [shard 0:sl:d] seastar_memory - oversized allocation: 3211264 bytes. This is non-fatal, but could lead to latency and/or fragmentation issues. Please report: at 0x269cfd9 0x269cb90 0x1bc0ab8 0x2876cd7 0x199bc94 0x420e364 0x18015b4 0x42153e2 0x1844bb5 0x4a2c8f9 0x180df86 0x18e6a52 0x18e1ad1 0x1855087 0x18767ff 0x412383a 0x15d0528 0x32fa458 0x32f9c69 0x32f96c3 0x32f7137 0x32f30c2 /lib64/libc.so.6+0x3574 /lib64/libc.so.6+0x3627 0x35a8904
...
ERROR 2025-11-13 11:56:36,526 [shard 0:sl:d] seastar_memory - Dumping seastar memory diagnostics
...
ERROR 2025-11-13 11:56:36,526 [shard 0:sl:d] seastar - Failed to allocate 131040 bytes
Aborting on shard 0, in scheduling group sl:default.
Backtrace:
```
The reason this failure affects other tests is that test.py reuses the same single-node Scylla cluster for all tests inside `test_secondary_index.py` (test.py runs tests file-by-file using a cluster that is leased from a pool). So, once a test causes the node to crash, all the tests that are defined after it inside the same file will fail as well.

I believe I didn't notice that in v1 because, at the time, the problematic test was the last one in the file. But later, some other PRs were merged that added more tests, and when I rebased, I wanted to keep closely related tests together, so I moved my new tests higher up in the file.

As for the crash, it's a known bug (https://github.com/scylladb/scylladb/issues/16295#issuecomment-3258819422). The difference in the logs is caused by `--abort-on-seastar-bad-alloc`.

**Comment by ndragazis** [comment]
v3 Changelog:
* Replaced `xfail` marker with `skip` for `test_short_read_static_column_map_values`.

**Comment by mykaul** [comment]
@nyh - can you please re-review?

File: test/cqlpy/test_secondary_index.py [hunk 1]
```diff
@@ -2063,44 +2063,67 @@ def test_index_in_API(cql, test_keyspace):
...
         # Test LIMIT within a single partition - succeeds.
-        rs = cql.execute(f'SELECT pk1, ck FROM {table} WHERE pk2 = 1 LIMIT 1')
-        assert sorted(list(rs)) == [(2,1)]
-        assert rs.has_more_pages == False
+        stmt = SimpleStatement(f'SELECT pk1, ck FROM {table} WHERE pk2 = 1 LIMIT 1', fetch_size=fetch_size)
+        rs = cql.execute(stmt)
+        if use_paging:
+            first_page = list(rs.current_rows)
+            assert sorted(first_page) == [(2,1)]
+            assert rs.has_more_pages == False
```
Comment (nyh) [inline test/cqlpy/test_secondary_index.py:2080]: Good catch about the rs.current_rows instead of reading all the pages. I don't understand how I ever saw has_more_pages = True in this test, strange.

File: test/cqlpy/test_secondary_index.py [hunk 2]
```diff
@@ -2063,44 +2063,67 @@ def test_index_in_API(cql, test_keyspace):
...
+        fetch_size = 100 if use_paging else None  # 100 is arbitrary; just large enough to exercise the paging path without splitting pages
         # Test LIMIT within a single partition slice - succeeds.
-        rs = cql.execute(f'SELECT pk, ck2 FROM {table} WHERE ck1 = 1 LIMIT 1')
-        assert sorted(list(rs)) == [(1,1)]
-        assert rs.has_more_pages == False
+        stmt = SimpleStatement(f'SELECT pk, ck2 FROM {table} WHERE ck1 = 1 LIMIT 1', fetch_size=fetch_size)
+        rs = cql.execute(stmt)
+        if use_paging:
+            first_page = list(rs.current_rows)
+            assert sorted(list(first_page)) == [(1,1)]
```
Comment (nyh) [inline test/cqlpy/test_secondary_index.py:2111]: no need for list(first_page) - it's already a list (done in the previous line).
You can also just do these two lines in one line instead of two:
```python
assert sorted(list(rs.current_rows)) = [(1,1)]
```

File: test/cqlpy/test_secondary_index.py [hunk 3]
```diff
@@ -2063,44 +2063,67 @@ def test_index_in_API(cql, test_keyspace):
...
+            assert sorted(list(rs)) == [(1,1)]
         # Test LIMIT across partition slices - reproduces #22158.
-        rs = cql.execute(f'SELECT pk, ck2 FROM {table} WHERE ck1 = 1 LIMIT 3')
-        assert sorted(list(rs)) == [(1,1), (1,2), (2,1)]
-        assert rs.has_more_pages == False
+        stmt = SimpleStatement(f'SELECT pk, ck2 FROM {table} WHERE ck1 = 1 LIMIT 3', fetch_size=fetch_size)
+        rs = cql.execute(stmt)
+        if use_paging:
+            first_page = list(rs.current_rows)
+            assert sorted(list(first_page)) == [(1,1), (1,2), (2,1)]
```
Comment (nyh) [inline test/cqlpy/test_secondary_index.py:2120]: again, first_page is already a list, and you can also just do this in one line, you don't need the first_page variable.

File: test/cqlpy/test_secondary_index.py [hunk 4]
```diff
@@ -2126,6 +2126,29 @@ def test_limit_partition_slice(cql, test_keyspace, use_paging):
             assert sorted(list(rs)) == [(1,1), (1,2), (2,1)]
 
 
+# Same as test_limit_partition_slice above, except that the indexed column is static.
+# Reproduces #22158.
+@pytest.mark.xfail(reason="issue #22158")
+@pytest.mark.parametrize("use_paging", [False, True])
```
Comment (nyh) [inline test/cqlpy/test_secondary_index.py:2128]: Why is the "with and without paging" distinction useful? Do you suspect the implementation has a different code path for the two cases?

File: query/query.cc [hunk 5]
```diff
@@ -362,10 +362,6 @@ static void write_partial_partition(ser::writer_of_qr_partition<bytes_ostream>&&
 }
 
 foreign_ptr<lw_shared_ptr<query::result>> result_merger::get() {
-    if (_partial.size() == 1) {
-        return std::move(_partial[0]);
-    }
```
Comment (nyh) [inline query/query.cc:350]: This change is slightly scary, because it changes some code which looks like it might be involved in all sorts of reads, not just from secondary indexes. Could it un-optimize (pessimize?) even single-item reads or that goes through a different code path?
Maybe you can run some sort of performance test (test/perf/...) before and after this patch to see nothing became worse?

Also would be nice to say in the commit message why this change is important - does it cause a specific bug? Is there some test that begins to pass after this change?

**Review by nyh — APPROVED** [review]
Sorry about the long delay in my review, I hope that as they say "better late than never"...

It's a very impressive patch. I'm not 100% sure I understood every bit of the logic, but it looks reasonable, and the testing - both existing ones and new ones - appears very careful and meticulous. Thanks.

I have one question about the change to the non-secondary index code, but I tentatively approve and will let you think if there's anything you want to change before I merge. In any case, please rebase your code on a new base so we'll get a new CI run - the existing one ran a long time ago and I don't know what might have changed in Scylla since.

Please also think what we should do with the issues that remain open but are mostly-solved - especially https://github.com/scylladb/scylladb/issues/25839. I propose that we should close (by "Fixes") this issue and open a new one explaining exactly what the remaining problem is. The new issue will also link to the old issue, and this pull request, and an earlier larger problem that was solved.

Also, if this patch also fixes https://github.com/scylladb/scylladb/issues/28026 (I'll check in a moment), please add Fixes of this issue too. Thanks!

**Comment by nyh** [comment]
https://github.com/scylladb/scylladb/pull/28108 was merged, so please rebase after it goes in and also un-xfail the new test test_short_count (which should pass - I verified with your PR that it does), and please also add Fixes for https://github.com/scylladb/scylladb/issues/28026. Thanks!
